### PR TITLE
Add informative error for infer batch size issues

### PR DESCRIPTION
**Pull request by hanlint**
This PR:
* adds an informative error if the batch yielded by the dataloader has a non-Tensor element (e.g. cannot retrieve the shape)
* improves test coverage of the `_default_num_samples_in_batch` function.

### CODE REVIEW

File: tests/trainer/test_dataspec.py [hunk 1]
```diff
@@ -0,0 +1,47 @@
...
+
+class TestDefaultGetNumSamples:
+
+    @pytest.fixture
+    def dataspec(self):
+        dataloader = DataLoader(RandomClassificationDataset())
+        return DataSpec(dataloader=dataloader)
+
+    # yapf: disable
+    @pytest.mark.parametrize('batch', [
```
Comment (dskhudia) [inline tests/trainer/test_dataspec.py:24]: Does it work on > 2-d tensors?

**Review by dskhudia — APPROVED** [review]

File: tests/trainer/test_dataspec.py (hunk 1 above)
Comment (hanlint) [reply tests/trainer/test_dataspec.py:24]: added a 4-dim tensor test!
